backendApi/src/main/java/com/virtuslab/gitmachete/backend/api/IGitMacheteRepositoryFactory.java

```diff
@@ -2,10 +2,21 @@
 
 import java.nio.file.Path;
 
+import io.vavr.collection.Set;
+import io.vavr.control.Option;
+
 import com.virtuslab.branchlayout.api.IBranchLayout;
 
 /** Each implementing class must have a public parameterless constructor. */
 public interface IGitMacheteRepositoryFactory {
-  IGitMacheteRepository create(Path mainDirectoryPath, Path gitDirectoryPath, IBranchLayout branchLayout)
-      throws GitMacheteException;
+  IGitMacheteRepository create(
+      Path mainDirectoryPath,
+      Path gitDirectoryPath,
+      IBranchLayout branchLayout) throws GitMacheteException;
+
+  Option<String> inferUpstreamForLocalBranch(
```

Comment:
This API sucks as for now... I hope it's gonna get better as I get more clarity on what's really needed when I write the discover feature

Reply:
yeah, probably something more precise than `String` could be returned here :thinking: